File: backend/src/test/java/ca/metricalsky/winston/entity/AuthorEntityTest.java
package ca.metricalsky.winston.entity;

import ca.metricalsky.winston.test.TestUtils;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;

import java.util.Set;

import static org.assertj.core.api.Assertions.assertThat;

@DataJpaTest
class AuthorEntityTest {

    @Autowired
    private TestEntityManager entityManager;

    @Test
    void persistsWithOnlyRequiredFields() {
        var authorEntity = AuthorEntity.builder()
                .id(TestUtils.randomId())
                .build();

        var persistedEntity = entityManager.persistFlushFind(authorEntity);

        assertThat(persistedEntity)
                .hasFieldOrPropertyWithValue("id", authorEntity.getId())
                .hasFieldOrPropertyWithValue("aliases", Set.of())
                .hasAllNullFieldsOrPropertiesExcept("aliases", "id", "lastFetchedAt");
    }

    @Test
    void persistsWithAllOptionalFields() {
        var authorAlias = TestUtils.randomString();
        var authorEntity = AuthorEntity.builder()
                .id(TestUtils.randomId())
                .displayName(TestUtils.randomString())
                .channelUrl(TestUtils.randomString())
                .profileImageUrl(TestUtils.randomString())
                .aliases(Set.of(authorAlias))
                .build();

        var persistedEntity = entityManager.persistFlushFind(authorEntity);

        assertThat(persistedEntity)
                .hasFieldOrPropertyWithValue("id", authorEntity.getId())
                .hasFieldOrPropertyWithValue("displayName", authorEntity.getDisplayName())
                .hasFieldOrPropertyWithValue("channelUrl", authorEntity.getChannelUrl())
                .hasFieldOrPropertyWithValue("profileImageUrl", authorEntity.getProfileImageUrl());
        assertThat(persistedEntity.getAliases())
                .containsExactly(authorAlias);
        assertThat(persistedEntity.getLastFetchedAt())
                .isNotNull();
    }
}

File: database/src/test/kotlin/ca/metricalsky/winston/database/entity/AuthorEntityTest.kt
package ca.metricalsky.winston.database.entity

import ca.metricalsky.winston.database.test.faker.DatabaseFaker
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull

@DataJpaTest
class AuthorEntityTest {

    val faker = DatabaseFaker()

    @Autowired
    lateinit var entityManager: TestEntityManager

    @Test
    fun `persists with only required fields`() {
        val author = faker.author().minimal()

        val persisted = entityManager.persistFlushFind(author);

        assertEquals(author.id, persisted.id)
        assertNull(persisted.displayName)
        assertNull(persisted.channelUrl)
        assertNull(persisted.profileImageUrl)
        assertNotNull(persisted.lastFetchedAt)
        assertEquals(setOf(), persisted.aliases)
    }

    @Test
    fun `persists with all optional fields`() {
        val author = faker.author().complete()

        val persisted = entityManager.persistFlushFind(author);

        assertEquals(author.id, persisted.id)
        assertEquals(author.displayName, persisted.displayName)
        assertEquals(author.channelUrl, persisted.channelUrl)
        assertEquals(author.profileImageUrl, persisted.profileImageUrl)
        assertNotNull(persisted.lastFetchedAt)
        assertEquals(author.aliases, persisted.aliases)
    }
}
